### PR TITLE
[FLINK-15646][K8s]Configurable K8s context support.

### PR DESCRIPTION
**Pull request by zhengcanbin**
## What is the purpose of the change

One would be forced to firstly use kubectl config use-context <context> to switch to the desired context if working with multiple K8S clusters or having multiple K8S "users" for interacting with the specified cluster,  so it is an important improvement to add an option(kubernetes.context) for configuring arbitrary contexts when deploying a Flink cluster. If that option is not specified, then the current context in the config file would be used.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhengcanbin** [comment]
cc @wangyang0918 @TisonKun

**Comment by wangyang0918** [comment]
@zhengcanbin The changes looks really good to me. I just have some minor comments.
1. The module in the commit title should be `k8s` or `kubernetes`. Maybe "[FLINK-15646][k8s] Make kubernetes context configurable" is better.
2. Even the abbreviation `k8s` is reasonable in the commit title to leave more characters for other useful information. I suggest to avoid in the code, documentation. This is not forced, and you could leave it as you want.

@TisonKun could you take another look and help to merge?

**Comment by zhengcanbin** [comment]
> @zhengcanbin The changes looks really good to me. I just have some minor comments.
> 
>     1. The module in the commit title should be `k8s` or `kubernetes`. Maybe "[FLINK-15646][k8s] Make kubernetes context configurable" is better.
> 
>     2. Even the abbreviation `k8s` is reasonable in the commit title to leave more characters for other useful information. I suggest to avoid in the code, documentation. This is not forced, and you could leave it as you want.
> 
k8s is more shorthand and used commonly in an informal context, sounds reasonable for the commit message.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java [hunk 1]
```diff
@@ -42,18 +42,23 @@ public static FlinkKubeClient fromConfiguration(Configuration flinkConfig) {
...
 			LOG.debug("Trying to load kubernetes config from file: {}.", kubeConfigFile);
 			try {
-				config = Config.fromKubeconfig(KubernetesUtils.getContentFromFile(kubeConfigFile));
+				config = Config.fromKubeconfig(kubeContext, KubernetesUtils.getContentFromFile(kubeConfigFile), null);
 			} catch (IOException e) {
 				throw new KubernetesClientException("Load kubernetes config failed.", e);
 			}
 		} else {
 			LOG.debug("Trying to load default kubernetes config.");
 			// Null means load from default context
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java]: We should update the comment here.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java [hunk 2]
```diff
@@ -42,18 +42,23 @@ public static FlinkKubeClient fromConfiguration(Configuration flinkConfig) {
 
 		final Config config;
 
+		final String kubeContext = flinkConfig.getString(KubernetesConfigOptions.CONTEXT);
+		if (kubeContext != null) {
+			LOG.info("Configuring K8S client using context {}.", kubeContext);
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java]: I agree to use "Kubernetes" in logging & document as @wangyang0918 suggested.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java [hunk 3]
```diff
@@ -29,6 +29,14 @@
 @PublicEvolving
 public class KubernetesConfigOptions {
 
+	public static final ConfigOption<String> CONTEXT =
+		key("kubernetes.context")
+		.stringType()
+		.noDefaultValue()
+		.withDescription("The desired context from your K8s config file used to configure the K8s client for " +
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java]: and here s/K8s/Kubernetes/

Comment (zhengcanbin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java]: Done

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java (hunk 2 above)
Comment (zhengcanbin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java]: Done

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java (hunk 1 above)
Comment (zhengcanbin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/KubeClientFactory.java]: Done

**Comment by zhengcanbin** [comment]
> Thanks for your contribution @zhengcanbin . I left some comments.
> 
> Beside, Azure fails on a known issue https://issues.apache.org/jira/browse/FLINK-15811. I'd suggest you rebase on nightly master and retest.

Thanks, @TisonKun ! Another commit was pushed to fix the comments.

**Pull request closed without merging**